### PR TITLE
RDK-38157 : Add MediaEngineRMF plugin

### PR DESCRIPTION
**Pull request by chaithents**
Reason for change: support for UVE-QAM playback.
Test Procedure: see ticket
Risks: low
Priority: P2
Signed-off-by: csojan <chaithents@tataelxsi.co.in>

### CODE REVIEW

File: MediaEngineRMF/MediaEngineRMF.json [hunk 1]
```diff
@@ -0,0 +1,9 @@
+{
```
Comment (anand-ky) [inline MediaEngineRMF/MediaEngineRMF.json:1]: MediaEngineRMF.json should define the API spec in json similar to other plugins. For instance https://github.com/rdkcentral/rdkservices/blob/sprint/22Q4/PersistentStore/PersistentStore.json

The current contents of this file should be moved to MediaEngineRMFPlugin.json

Comment (chaithents) [reply MediaEngineRMF/MediaEngineRMF.json:1]: @anand-ky Will fix the JSON file. Thanks for the pointers.

File: MediaEngineRMF/MediaEngineRMF.h [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+            static const string EVT_ON_STATUS;
+            static const string EVT_ON_ERROR;
+            static const string EVT_ON_WARNING;
+
+
+        private/*registered methods*/:
+            //methods
+            uint32_t initialize_player(const JsonObject& parameters, JsonObject& response);
+            uint32_t deinitialize_player(const JsonObject& parameters, JsonObject& response);
+            uint32_t play(const JsonObject& parameters, JsonObject& response);
```
[the remark this answers is not in the document]
Comment (chaithents) [reply MediaEngineRMF/MediaEngineRMF.h:86]: @anand-ky The plugin is directly linked with a library (libmediaplayer.so) that does the heavy lifting. That library is the native app right now. We may explore changing this to an out-of-process plugin at some point. Can the COM RPC interface wait till then?

**Comment by npoltorapavlo** [comment]
fyi it's no longer XRE SERVICE MANAGER project . please follow JSON RPC spec for input and output objects of the JSON RPC interface and use a repo with interfaces https://github.com/rdkcentral/ThunderInterfaces

File: MediaEngineRMF/MediaEngineRMF.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (anand-ky) [reply MediaEngineRMF/MediaEngineRMF.h:86]: sounds good. Do we have some data on CPU/RAM usage of WPEFramework process during QAM playback.

**Comment by anand-ky** [comment]
> > Can we also look at adding unit tests for this plugin using google unit tests. More details at https://github.com/rdkcentral/rdkservices/tree/sprint/22Q4/Tests
> 
> Certainly. However, this user story is a 2022 Q4 deliverable. I don't think I can finish the unit tests as well within that time frame. Can we add this later?

yes the unit tests can come later.

File: MediaEngineRMF/MediaEngineRMF.json (hunk 1 above)
Comment (chaithents) [reply MediaEngineRMF/MediaEngineRMF.json:1]: @anand-ky Updated JSON API spec and renamed/moved files as you suggested. Can you please check?

**Comment by chaithents** [comment]
> fyi it's no longer XRE SERVICE MANAGER project . please follow JSON RPC spec for input and output objects of the JSON RPC interface and use a repo with interfaces https://github.com/rdkcentral/ThunderInterfaces

@npoltorapavlo I went through https://github.com/rdkcentral/ThunderInterfaces and found that most of the plugins there carry a Metrological banner, and many prominent plugins (SystemServices, MaintenanceManager etc.) closely tied to Comcast flavour RDK are missing. MediaEngineRMF leans in the Comcast-RDK direction, so I'm a little confused. Will check further and get back to you.

File: MediaEngineRMF/MediaEngineRMF.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (anand-ky) [reply MediaEngineRMF/MediaEngineRMF.h:86]: sounds good, thanks.

**Comment by anand-ky** [comment]
> > fyi it's no longer XRE SERVICE MANAGER project . please follow JSON RPC spec for input and output objects of the JSON RPC interface and use a repo with interfaces https://github.com/rdkcentral/ThunderInterfaces
> 
> @npoltorapavlo I went through https://github.com/rdkcentral/ThunderInterfaces and found that most of the plugins there carry a Metrological banner, and many prominent plugins (SystemServices, MaintenanceManager etc.) closely tied to Comcast flavour RDK are missing. MediaEngineRMF leans in the Comcast-RDK direction, so I'm a little confused. Will check further and get back to you.

You can check how it's done for PersistentStore. ThunderInterfaces host the C++ interfaces.
https://github.com/rdkcentral/rdkservices/pull/2287/files
https://github.com/rdkcentral/ThunderInterfaces/pull/127/files

**Comment by npoltorapavlo** [comment]
> > > fyi it's no longer XRE SERVICE MANAGER project . please follow JSON RPC spec for input and output objects of the JSON RPC interface and use a repo with interfaces https://github.com/rdkcentral/ThunderInterfaces
> > 
> > 
> > @npoltorapavlo I went through https://github.com/rdkcentral/ThunderInterfaces and found that most of the plugins there carry a Metrological banner, and many prominent plugins (SystemServices, MaintenanceManager etc.) closely tied to Comcast flavour RDK are missing. MediaEngineRMF leans in the Comcast-RDK direction, so I'm a little confused. Will check further and get back to you.
> 
> You can check how it's done for PersistentStore. ThunderInterfaces host the C++ interfaces. https://github.com/rdkcentral/rdkservices/pull/2287/files https://github.com/rdkcentral/ThunderInterfaces/pull/127/files

my comment was about JSON RPC interface
e.g
https://github.com/rdkcentral/ThunderInterfaces/blob/master/jsonrpc/Messenger.json - interface.

https://github.com/rdkcentral/rdkservices/blob/sprint/22Q4/Messenger/MessengerJsonRpc.cpp
includes generated header `#include <interfaces/json/JsonData_Messenger.h>` , uses generated input and output objects `uint32_t Messenger::endpoint_join(const JoinParamsData& params, JoinResultInfo& response)`, returns error codes as per JSON RPC spec `return result? Core::ERROR_NONE : Core::ERROR_UNKNOWN_KEY;`

**Comment by anand-ky** [comment]
> > > > fyi it's no longer XRE SERVICE MANAGER project . please follow JSON RPC spec for input and output objects of the JSON RPC interface and use a repo with interfaces https://github.com/rdkcentral/ThunderInterfaces
> > > 
> > > 
> > > @npoltorapavlo I went through https://github.com/rdkcentral/ThunderInterfaces and found that most of the plugins there carry a Metrological banner, and many prominent plugins (SystemServices, MaintenanceManager etc.) closely tied to Comcast flavour RDK are missing. MediaEngineRMF leans in the Comcast-RDK direction, so I'm a little confused. Will check further and get back to you.
> > 
> > 
> > You can check how it's done for PersistentStore. ThunderInterfaces host the C++ interfaces. https://github.com/rdkcentral/rdkservices/pull/2287/files https://github.com/rdkcentral/ThunderInterfaces/pull/127/files
> 
> my comment was about JSON RPC interface e.g https://github.com/rdkcentral/ThunderInterfaces/blob/master/jsonrpc/Messenger.json - interface.
> 
> https://github.com/rdkcentral/rdkservices/blob/sprint/22Q4/Messenger/MessengerJsonRpc.cpp includes generated header `#include <interfaces/json/JsonData_Messenger.h>` , uses generated input and output objects `uint32_t Messenger::endpoint_join(const JoinParamsData& params, JoinResultInfo& response)`, returns error codes as per JSON RPC spec `return result? Core::ERROR_NONE : Core::ERROR_UNKNOWN_KEY;`

@chaithents - Merging this PR to meet sprint deadlines. Pls bring the other changes as Nikita detailed in a follow-up ticket.

**Review by anand-ky — APPROVED** [review]

**Pull request merged**